test/test_cuda.py

```diff
@@ -3967,6 +3968,75 @@ def test_lazy_init(self):
         rc = check_output([sys.executable, '-c', test_script]).decode("ascii").strip()
         self.assertEqual(rc, "0")
 
+    def test_event_creation_callback(self):
+        torch._C._activate_cuda_trace()
+        self.mock = unittest.mock.MagicMock()
+        cuda_trace.register_callback_for_cuda_event_creation(self.mock)
+
+        event = torch.cuda.Event()
+        event.record()
+        self.mock.assert_called_once_with(event._as_parameter_.value)
+
+    def test_event_deletion_callback(self):
+        torch._C._activate_cuda_trace()
+        self.mock = unittest.mock.MagicMock()
+        cuda_trace.register_callback_for_cuda_event_deletion(self.mock)
+
+        event = torch.cuda.Event()
+        event.record()
+        event_id = event._as_parameter_.value
+        del event
+        self.mock.assert_called_once_with(event_id)
+
+    def test_event_record_callback(self):
+        torch._C._activate_cuda_trace()
+        self.mock = unittest.mock.MagicMock()
+        cuda_trace.register_callback_for_cuda_event_record(self.mock)
+
+        event = torch.cuda.Event()
+        event.record()
+        self.mock.assert_called_once_with(
+            event._as_parameter_.value, torch.cuda.default_stream().cuda_stream
+        )
+
+    def test_event_wait_callback(self):
+        torch._C._activate_cuda_trace()
+        self.mock = unittest.mock.MagicMock()
+        cuda_trace.register_callback_for_cuda_event_wait(self.mock)
+
+        event = torch.cuda.Event()
+        event.record()
+        event.wait()
+        self.mock.assert_called_once_with(
+            event._as_parameter_.value, torch.cuda.default_stream().cuda_stream
+        )
+
+    def test_memory_allocation_callback(self):
+        torch._C._activate_cuda_trace()
+        self.mock = unittest.mock.MagicMock()
+        cuda_trace.register_callback_for_cuda_memory_allocation(self.mock)
+
+        tensor = torch.empty(10, 4, device="cuda")
+        self.mock.assert_called_once_with(tensor.data_ptr())
+
+    def test_memory_deallocation_callback(self):
+        torch._C._activate_cuda_trace()
+        self.mock = unittest.mock.MagicMock()
+        cuda_trace.register_callback_for_cuda_memory_deallocation(self.mock)
+
+        tensor = torch.empty(3, 8, device="cuda")
+        data_ptr = tensor.data_ptr()
+        del tensor
+        self.mock.assert_called_once_with(data_ptr)
+
+    def test_stream_creation_callback(self):
```

Comment:
I realize that this is a tricky one, right? Because the entire pool of streams is initialized in full, lazily, when first needed. I am thus also a bit worried about what happens if this test is run after another test and this one already initializes the streams: in this case the init in this test would be a no-op and the hook wouldn't be invoked.